cachi2/core/package_managers/npm.py

```diff
@@ -0,0 +1,226 @@
+        """Return a list of dicts with sbom component data."""
+        packages = self._dependencies if self.lockfile_version == 1 else self._packages
+        return [{"name": package.name, "version": package.version} for package in packages]
+
+
+def fetch_npm_source(request: Request) -> RequestOutput:
+    """Resolve and fetch npm dependencies for the given request.
+
+    :param request: the request to process
+    :return: A RequestOutput object with content for all npm packages in the request
+    """
+    components: List[Component] = []
+    project_files: List[ProjectFile] = []
+
+    for package in request.npm_packages:
+        info = _resolve_npm(
+            request.source_dir.join_within_root(package.path),
+            request.output_dir,
+        )
+
+        components.append(Component.from_package_dict(info["package"]))
+        project_files.append(info["package_lock_file"])
+
+        for dependency in info["dependencies"]:
+            components.append(Component.from_package_dict(dependency))
+
+    return RequestOutput.from_obj_list(
+        components=components,
+        environment_variables=[],
+        project_files=project_files,
+    )
+
+
+def _resolve_npm(pkg_path: RootedPath, output_dir: RootedPath) -> Dict[str, Any]:
+    """Resolve and fetch npm dependencies for the given package.
+
+    :param pkg_path: the path to the directory containing npm-shrinkwrap.json or package-lock.json
+    :param output_dir: the path to the directory containing npm-shrinkwrap.json or package-lock.json
+    :return: a dictionary that has the following keys:
+        ``package`` which is the dict representing the main Package,
+        ``dependencies`` which is a list of dicts representing the package Dependencies
+        ``package_lock_file`` which is the (updated) package-lock.json as a ProjectFile
+    :raises PackageRejected: if the npm package is not cachi2 compatible
+    """
+    # npm-shrinkwrap.json and package-lock.json share the same format but serve slightly
+    # different purposes. See the following documentation for more information:
+    # https://docs.npmjs.com/files/package-lock.json.
+    for lock_file in ("npm-shrinkwrap.json", "package-lock.json"):
+        package_lock_path = pkg_path.join_within_root(lock_file)
+        if package_lock_path.path.exists():
+            break
+    else:
+        raise PackageRejected(
+            "The npm-shrinkwrap.json or package-lock.json file must be present for the npm "
+            "package manager",
+            solution="Please double-check that you have specified the correct path to the package directory containing one of those two files",
+        )
+
+    package_lock = PackageLock.from_file(package_lock_path, output_dir)
+
+    return {
+        "package": package_lock.main_package,
+        "dependencies": package_lock.get_sbom_components(),
+        "package_lock_file": package_lock.get_project_file(),
+    }
```

Comment:
Could we return something typed? The example set by gomod and pip is one that we definitely shouldn't follow :sweat_smile:

TypedDict, NamedTuple, dataclass, anything other than a `dict[str, Any]`

(TypedDict should be the fewest changes, just the return annotation)